databuilder/query_engines/sqlite.py

```diff
@@ -12,12 +15,93 @@ class SQLiteQueryEngine(BaseQueryEngine):
-            """
+        column_expressions = {
+            name: self.get_sql(definition)
+            for name, definition in column_definitions.items()
+        }
+        population_expression = column_expressions.pop("population")
+        return self.get_results_query(column_expressions, population_expression)
+
+    def get_results_query(self, column_expressions, population_expression):
+        # Get all referenced tables
+        population_tables = set(get_tables(population_expression))
+        other_tables = {
+            table for expr in column_expressions.values() for table in get_tables(expr)
+        }
+        all_tables = population_tables | other_tables
+
+        # Build a column expression for `patient_id`. In the simplest case there's a
+        # single table in the query which is guaranteed to have rows for every patient
+        # and so we can use that. But the flexibility of the query model means that we
+        # need to cope with more complex situations. Because of the outer joins between
+        # tables there may be no one table which has a row for every patient in the
+        # query.
+
+        if len(population_tables) > 0:  # pragma: no cover
+            # If the population expression contains table references (which it might not
+            # because it could be the literal value `True`) then those tables together
+            # must contain all the patient_ids we need
+            patient_id_tables = population_tables
+        else:
+            # Otherwise, we have to check all the tables
+            patient_id_tables = all_tables
+
+        if len(patient_id_tables) == 1:
+            # If there's only one table then our patient_id expression is just the
+            # patient_id column from this table
+            patient_id = list(patient_id_tables)[0].c.patient_id
+        elif len(patient_id_tables) > 1:  # pragma: no cover
+            # Otherwise we have to coalesce the patient_ids from each of the source
+            # tables to pick the first non-null value
+            patient_id = sqlalchemy.func.coalesce(
+                *[table.c.patient_id for table in patient_id_tables]
+            )
+        else:
+            assert False, "No tables found in query"
+
+        joins = None
+        for table in all_tables:
+            if joins is None:
+                joins = table
+            else:
+                joins = joins.join(
+                    table, table.c.patient_id == patient_id, isouter=True
+                )  # pragma: no cover
+
+        columns = [
+            patient_id.label("patient_id"),
+            *[expr.label(name) for name, expr in column_expressions.items()],
+        ]
+
+        return (
+            sqlalchemy.select(columns).select_from(joins).where(population_expression)
         )
 
+    @singledispatchmethod
+    def get_sql(self, node):
+        assert False, f"Unhandled node: {node}"
```

Comment:
I'd love to have these all as functions too, but I'm not sure it's going to be easily possible because (although this isn't yet the case) it's not just the table mapping for which these methods will need `self`. There are all sorts of places where we need to specialise behaviour per query-engine (e.g. how to create a temporary table) in the middle of walking the graph and I think the most practical way of doing that is to have these as methods on a base class.

> It might be easier to understand this code if we clarify that by having a table mapping object that it can return rather than calling a function on it for each table. This then becomes clearly data.

The reason I liked the idea of using functions here was that I wanted to leave open the possibility of the query engine giving the backend more information about the kind of query it's going to do (e.g these are all the columns I'll need) and letting the backend optimise the kind of query it returns. In my head was something like SQLite's [Virtual Tables](https://www.sqlite.org/vtab.html). I think this stems from worries I had when first designing all this stuff that it would generate awful performance problems and this was my way of mentally waving those away! But we're not doing any that stuff; we've got not immediate plans to; and if we decide we need to do it then we can change the API. So I'd be happy to change it to something declarative if you think that would be an improvement.

> The backend is quite a rich object and we only need a small part of it here.

Yes, I very much agree with this and I've been thinking about change the QueryEngine API for a while, but this didn't feel like the right PR. In my head I had something like this:
```py
query_engine = QueryEngine(database_url, get_table_expression=backend.get_table_expression)
results = query_enginge.fetch_results(dataset_definition)
```
That is, just passing the bit of the backend we care about to the constructor, and not passing the dataset definition itself but supplying that later.

I'm slightly in two minds about passing the database_url to the constructor as well. Should you _have_ to have a database to be able to construct a query engine? Maybe it should be a parameter to `fetch_results()`.

It's worth getting right I think, but also easy enough to change later so maybe doesn't need to much thought right now.